README.md
go-camo
=======

## About

Go version of [Camo][1] server.

[Camo][1] is a special type of image proxy that proxies non-secure images over
SSL/TLS. This prevents mixed content warnings on secure pages.

It works in conjunction with back-end code to rewrite image URLs and sign them
with an [HMAC][4].

## How it works

First you parse the original URL, generate an HMAC signature of it, then hex
encode it, and then place the pieces into the expected format replacing the
original image URL.

The client requests the URL to Go-Camo. Go-Camo validates the HMAC, decodes the
URL, requests the content and streams it to the client.

Here is some example python code that demonstrates generating an encoded URL:

    import hashlib
    import hmac
    def mk_camo_url(hmac_key, image_url, camo_host):
        if image_url.startswith("https:"):
            return image_url
        hexdigest = hmac.new(hmac_key, image_url, hashlib.sha1).hexdigest()
        hexurl = image_url.encode('hex')
        requrl = 'https://%s/%s/%s' % (camo_host, hexdigest, hexurl)
        return requrl

Here it is in action:

    >>> mk_camo_url("test", "http://golang.org/doc/gopher/frontpage.png", "img.example.org")
    'https://img.example.org/0f6def1cb147b0e84f39cbddc5ea10c80253a6f3/687474703a2f2f676f6c616e672e6f72672f646f632f676f706865722f66726f6e74706167652e706e67'

While Go-Camo will support proxying HTTPS images as well, for performance
reasons you may choose to filter HTTPS requests out from proxying, and let the
client simply fetch those as they are. The code example above does this.

Note that it is recommended to front Go-Camo with a CDN when possible.

## Differences from Camo

*   Go-Camo supports 'Path Format' url format only. Camo's "Query
    String Format" is not supported.
*   Go-Camo supports "allow regex host filters".
*   Go-Camo supports client http keep-alives.
*   Go-Camo provides native SSL support.
*   Go-Camo supports using more than one os thread (via GOMAXPROCS) without the
    need of multiple instances or additional proxying.
*   Go-Camo builds to a static binary. This makes deploying to large numbers
    of servers a snap.
*   Go-Camo supports a Base64 url encoding option (goal is reduction of url
    size).

## Building

Building requires `git` and `hg` (mecurial). They are used to fetch
dependencies. A functional [Go][3] installation is also required.

    # Set GOPATH if appropriate

    # get code and dependencies
    $ go get -d github.com/cactus/go-camo

    # build and install to GOPATH
    $ go install github.com/cactus/go-camo

    # as an alternative to the previous command, build and strip debug symbols.
    # this is useful for production, and reduces the resulting file size.
    $ go install -ldflags '-s' github.com/cactus/go-camo


## Running

    $ $GOPATH/bin/go-camo -c config.json

Go-Camo does not daemonize on its own. For production usage, it is recommended
to launch in a process supervisor, and drop privileges as appropriate.

Examples of supervisors include: [daemontools][5], [runit][6], [upstart][7],
[launchd][8], and many more.

For the reasoning behind lack of daemonization, see [daemontools/why][9]. In
addition, the code is much simpler because of it.

## Running under devweb

[Devweb][2] is useful for developing. To run under devweb:

    $ go get code.google.com/p/rsc/devweb
    $ PATH=.:$PATH $GOPATH/bin/devweb -addr=127.0.0.1:8080 github.com/cactus/go-camo/go-camo-devweb
    $ rm -f ./prox.exe  # devweb drops this file. clean it up

## Configuring

    $ $GOPATH/bin/go-camo -h
    Usage:
      go-camo [OPTIONS]

    Help Options:
      -h, --help          Show this help message

    Application Options:
      -c, --config        JSON Config File
      -k, --key           HMAC key
          --stats         Enable Stats
          --max-size      Max response image size (KB) (5120)
          --timeout       Upstream request timeout (4s)
          --no-follow     Disable following upstream redirects
          --listen        Address:Port to bind to for HTTP (0.0.0.0:8080)
          --ssl-listen    Address:Port to bind to for HTTPS/SSL/TLS
          --ssl-key       ssl private key (key.pem) path
          --ssl-cert      ssl cert (cert.pem) path
      -v, --verbose       Show verbose (debug) log level output
      -V, --version       print version and exit

    $ cat config.json
    {
        "HmacKey": "Some long string here...",
        "AllowList": []
    }

*   `HmacKey` is a secret key seed to the HMAC used for signing and
    validation.
*   `Allowlist` is a list of regex host matches to allow.

If an AllowList is defined, and a request does not match one of the listed host
regex, then the request is denied. Default is all requests pass the Allowlist
if none is specified.

Option flags, if provided, override those in the config file.

If stats flag is provided, then the service will track bytes and clients
served, and offer them up at an http endpoint `/status` via HTPT GET request.

## Additional tools

Go-Camo includes a couple of additional tools.

### url-tool

The `url-tool` utility provides a simple way to generate signed URLs from the command line.

    $ $GOPATH/bin/url-tool -h
    Usage:
      url-tool [OPTIONS]

    Help Options:
      -h, --help      Show this help message

    Application Options:
      -c, --config    JSON Config File
      -k, --key       HMAC key
      -e, --encode    Encode a url and print result
      -d, --decode    Decode a url and print result
          --prefix    Optional url prefix used by encode output

Example usage:

    $ $GOPATH/bin/url-tool -e -k "test" --prefix="https://img.example.org" "http://golang.org/doc/gopher/frontpage.png"
    https://img.example.org/0f6def1cb147b0e84f39cbddc5ea10c80253a6f3/687474703a2f2f676f6c616e672e6f72672f646f632f676f706865722f66726f6e74706167652e706e67 

Installation:

    $ go install github.com/cactus/go-camo/url-tool

### simple-server

The `simple-server` utility is useful for testing. It serves the contents of a
given directory over http. Nothing more.

    $ $GOPATH/bin/simple-server -h
    Usage of ./simple-server:
      -d=".": Directory to serve from

Installation:

    $ go install github.com/cactus/go-camo/simple-server

## Changelog

See `CHANGELOG.md`

## License

Released under the [MIT
license](http://www.opensource.org/licenses/mit-license.php). See `LICENSE.md`
file for details.

[1]: https://github.com/atmos/camo
[2]: http://code.google.com/p/rsc/source/browse/devweb
[3]: http://golang.org/doc/install
[4]: http://en.wikipedia.org/wiki/HMAC
[5]: http://cr.yp.to/daemontools.html
[6]: http://smarden.org/runit/
[7]: http://upstart.ubuntu.com/
[8]: http://launchd.macosforge.org/
[9]: http://cr.yp.to/daemontools/faq/create.html#why


CHANGELOG.md
Changelog
=========

## 0.1.0 in-progress

*   Refactor logging a bit
*   Move encoding functionality into a submodule to reduce import size (and
    thus resultant binary size) for url-tool
*   Prevent request loop
*   Remove custom Denylist support. Filtering should be done on signed url
    generation. rfc1918 filtering retained and internalized so as do reduce
    internal network exposue surface and avoid non-routable effort.
*   Inverted redirect boolean. Redirects are now followed by default, and 
    the flag `no-follow` was learned.
*   Use new flag parsing library for nicer help and cleaner usage.
*   Support Base64 url encoding option

## 0.0.4 2012-09-02

*   Refactor Stats code out of camoproxy
*   Make stats an optional flag in go-camo
*   Minor documentation cleanup
*   Clean up excessive logging on client initiated broken pipe

## 0.0.3 2012-08-05

*   organize and clean up code
*   make header filters exported 
*   start filtering response headers
*   add default Server name
*   fix bug dealing with header filtering logic
*   add cli utility to encode/decode urls for testing, etc.
*   change repo layout to be friendlier for Go development/building
*   timeout flag is now a duration (15s, 10m, 1h, etc)
*   X-Forwarded-For support
*   Added more info to readme


## 0.0.2 2012-07-12

*   documentation cleanup
*   code reorganization
*   some cleanup of command flag text
*   logging code simplification


## 0.0.1 2012-07-07

*   initial release


url-tool/main.go
// go-camo daemon (go-camod)
package main

import (
	"encoding/json"
	"fmt"
	"github.com/cactus/go-camo/camoproxy/encoding"
	flags "github.com/jessevdk/go-flags"
	"io/ioutil"
	"log"
	"net/url"
	"os"
	"strings"
)

func main() {

	// command line flags
	var opts struct {
		ConfigFile string `short:"c" long:"config" description:"JSON Config File"`
		HmacKey    string `short:"k" long:"key" description:"HMAC key"`
		Encode     bool   `short:"e" long:"encode" description:"Encode a url and print result"`
		Decode     bool   `short:"d" long:"decode" description:"Decode a url and print result"`
		Base64     bool   `long:"base64" description:"Use Base64 encoding for url. Only relevant for Encode."`
		Hex        bool   `long:"hex" description:"Use hex encoding for url. Only relevant for Encode."`
		Prefix     string `long:"prefix" default:"" description:"Optional url prefix used by encode output"`
	}

	// parse said flags
	args, err := flags.Parse(&opts)
	if err != nil {
		if e, ok := err.(*flags.Error); ok {
			if e.Type == flags.ErrHelp {
				os.Exit(0)
			}
		}
		os.Exit(1)
	}

	// clear log prefix -- not needed for tool
	log.SetFlags(0)

	// Anonymous struct Container for holding configuration parameters
	// parsed from JSON config file.
	config := struct {
		HmacKey string
	}{}

	if opts.ConfigFile != "" {
		b, err := ioutil.ReadFile(opts.ConfigFile)
		if err != nil {
			log.Fatal("Could not read configFile", err)
		}
		err = json.Unmarshal(b, &config)
		if err != nil {
			log.Fatal("Could not parse configFile", err)
		}
	}

	if opts.Encode == true && opts.Decode == true {
		log.Fatal("Encode and Decode are mutually exclusive. Doing nothing.")
	}

	if opts.Encode == false && opts.Decode == false {
		log.Fatal("No action requested. Doing nothing.")
	}

	// flags override config file
	if opts.HmacKey != "" {
		config.HmacKey = opts.HmacKey
	}

	if len(args) == 0 {
		log.Fatal("No url argument provided")
	}

	oUrl := args[0]
	if oUrl == "" {
		log.Fatal("No url argument provided")
	}

	hmacKeyBytes := []byte(config.HmacKey)

	var encoder encoding.Encoder
	var decoder encoding.Decoder
	var suffix string
	if opts.Base64 {
		encoder = encoding.EncodeBase64Url
		decoder = encoding.DecodeBase64Url
		suffix = "?e=base64"
	}
	if !opts.Base64 || opts.Hex {
		encoder = encoding.EncodeHexUrl
		decoder = encoding.DecodeHexUrl
		suffix = ""
	}

	if opts.Encode == true {
		outUrl := encoder(&hmacKeyBytes, oUrl)
		fmt.Println(opts.Prefix + outUrl + suffix)
	}

	if opts.Decode == true {
		u, err := url.Parse(oUrl)
		if err != nil {
			log.Fatal(err)
		}
		qs := u.Query()
		if qs.Get("e") == "base64" {
			decoder = encoding.DecodeBase64Url
		} else {
			decoder = encoding.DecodeHexUrl
		}
		comp := strings.SplitN(u.Path, "/", 3)
		decUrl, valid := decoder(&hmacKeyBytes, comp[1], comp[2])
		if !valid {
			log.Fatal("hmac is invalid")
		}
		log.Println(decUrl)
	}
}


camoproxy/proxy.go
// Package camoproxy provides an HTTP proxy server with content type
// restrictions as well as regex host allow list support.
package camoproxy

import (
	"code.google.com/p/gorilla/mux"
	"errors"
	"github.com/cactus/go-camo/camoproxy/encoding"
	"github.com/cactus/gologit"
	"io"
	"net"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"syscall"
	"time"
)

// Config holds configuration data used when creating a Proxy with New.
type Config struct {
	// HmacKey is a string to be used as the hmac key
	HmacKey           string
	// AllowList is a list of string represenstations of regex (not compiled
	// regex) that are used as a whitelist filter. If an AllowList is present,
	// then anything not matching is dropped. If no AllowList is present,
	// no Allow filtering is done.
	AllowList         []string
	// MaxSize is the maximum valid image size response (in bytes).
	MaxSize           int64
	// NoFollowRedirects is a boolean that specifies whether upstream redirects
	// are followed (10 depth) or not.
	NoFollowRedirects bool
	// Request timeout is a timeout for fetching upstream data.
	RequestTimeout    time.Duration
}

// Interface for Proxy to use for stats/metrics.
// This must be goroutine safe, as AddBytes and AddServed will be called from
// many goroutines.
type ProxyMetrics interface {
	AddBytes(bc int64)
	AddServed()
}

// A Proxy is a Camo like HTTP proxy, that provides content type
// restrictions as well as regex host allow list support.
type Proxy struct {
	client    *http.Client
	hmacKey   []byte
	allowList []*regexp.Regexp
	maxSize   int64
	metrics   ProxyMetrics
}

// ServeHTTP handles the client request, validates the request is validly
// HMAC signed, filters based on the Allow list, and then proxies
// valid requests to the desired endpoint. Responses are filtered for
// proper image content types.
func (p *Proxy) ServeHTTP(w http.ResponseWriter, req *http.Request) {
	gologit.Debugln("Request:", req.URL)
	if p.metrics != nil {
		go p.metrics.AddServed()
	}

	var decoder encoding.Decoder
	if req.URL.Query().Get("e") == "base64" {
		decoder = encoding.DecodeBase64Url
	} else {
		decoder = encoding.DecodeHexUrl
	}

	w.Header().Set("Server", ServerNameVer)

	vars := mux.Vars(req)
	surl, ok := decoder(&p.hmacKey, vars["sigHash"], vars["encodedUrl"])
	if !ok {
		http.Error(w, "Bad Signature", http.StatusForbidden)
		return
	}
	gologit.Debugln("URL:", surl)

	u, err := url.Parse(surl)
	if err != nil {
		gologit.Debugln(err)
		http.Error(w, "Bad url", http.StatusBadRequest)
		return
	}

	u.Host = strings.ToLower(u.Host)
	if u.Host == "" || localhostRegex.MatchString(u.Host) {
		http.Error(w, "Bad url", http.StatusNotFound)
		return
	}

	if req.Header.Get("Via") == "ServerNameVer" {
		http.Error(w, "Request loop failure", http.StatusNotFound)
	}

	// if allowList is set, require match
	matchFound := true
	if len(p.allowList) > 0 {
		matchFound = false
		for _, rgx := range p.allowList {
			if rgx.MatchString(u.Host) {
				matchFound = true
			}
		}
	}
	if !matchFound {
		http.Error(w, "Allowlist host failure", http.StatusNotFound)
		return
	}

	// filter out rfc1918 hosts
	ip := net.ParseIP(u.Host)
	if ip != nil {
		if addr1918PrefixRegex.MatchString(ip.String()) {
			http.Error(w, "Denylist host failure", http.StatusNotFound)
			return
		}
	}

	nreq, err := http.NewRequest("GET", surl, nil)
	if err != nil {
		gologit.Debugln("Could not create NewRequest", err)
		http.Error(w, "Error Fetching Resource", http.StatusBadGateway)
		return
	}

	// filter headers
	p.copyHeader(&nreq.Header, &req.Header, &ValidReqHeaders)
	if req.Header.Get("X-Forwarded-For") == "" {
		host, _, err := net.SplitHostPort(req.RemoteAddr)
		if err == nil && !addr1918PrefixRegex.MatchString(host) {
			nreq.Header.Add("X-Forwarded-For", host)
		}
	}
	nreq.Header.Add("connection", "close")
	nreq.Header.Add("user-agent", ServerNameVer)
	nreq.Header.Add("via", ServerNameVer)

	resp, err := p.client.Do(nreq)
	if err != nil {
		gologit.Debugln("Could not connect to endpoint", err)
		if strings.Contains(err.Error(), "timeout") {
			http.Error(w, "Error Fetching Resource", http.StatusBadGateway)
		} else {
			http.Error(w, "Error Fetching Resource", http.StatusNotFound)
		}
		return
	}
	defer resp.Body.Close()

	// check for too large a response
	if resp.ContentLength > p.maxSize {
		gologit.Debugln("Content length exceeded", surl)
		http.Error(w, "Content length exceeded", http.StatusNotFound)
		return
	}

	switch resp.StatusCode {
	case 200:
		// check content type
		ct, ok := resp.Header[http.CanonicalHeaderKey("content-type")]
		if !ok || ct[0][:6] != "image/" {
			gologit.Debugln("Non-Image content-type returned", u)
			http.Error(w, "Non-Image content-type returned",
				http.StatusBadRequest)
			return
		}
	case 300:
		gologit.Debugln("Multiple choices not supported")
		http.Error(w, "Multiple choices not supported", http.StatusNotFound)
		return
	case 301, 302, 303, 307:
		// if we get a redirect here, we either disabled following,
		// or followed until max depth and still got one (redirect loop)
		http.Error(w, "Not Found", http.StatusNotFound)
		return
	case 304:
		h := w.Header()
		p.copyHeader(&h, &resp.Header, &ValidRespHeaders)
		h.Set("X-Content-Type-Options", "nosniff")
		w.WriteHeader(304)
		return
	case 404:
		http.Error(w, "Not Found", http.StatusNotFound)
		return
	case 500, 502, 503, 504:
		// upstream errors should probably just 502. client can try later.
		http.Error(w, "Error Fetching Resource", http.StatusBadGateway)
		return
	default:
		http.Error(w, "Not Found", http.StatusNotFound)
		return
	}

	h := w.Header()
	p.copyHeader(&h, &resp.Header, &ValidRespHeaders)
	h.Set("X-Content-Type-Options", "nosniff")
	w.WriteHeader(resp.StatusCode)

	// since this uses io.Copy from the respBody, it is streaming
	// from the request to the response. This means it will nearly
	// always end up with a chunked response.
	// Change to the following to send whole body at once, and
	// read whole body at once too:
	//    body, err := ioutil.ReadAll(resp.Body)
	//    if err != nil {
	//        gologit.Println("Error writing response:", err)
	//    }
	//    w.Write(body)
	// Might use quite a bit of memory though. Untested.
	bW, err := io.Copy(w, resp.Body)
	if err != nil {
		// only log if not broken pipe. broken pipe means the client
		// terminated conn for some reason.
		opErr, ok := err.(*net.OpError)
		if !ok || opErr.Err != syscall.EPIPE {
			gologit.Println("Error writing response:", err)
		}
		return
	}

	if p.metrics != nil {
		go p.metrics.AddBytes(bW)
	}
	gologit.Debugln(req, resp.StatusCode)
}

// copy headers from src into dst
// empty filter map will result in no filtering being done
func (p *Proxy) copyHeader(dst, src *http.Header, filter *map[string]bool) {
	f := *filter
	filtering := false
	if len(f) > 0 {
		filtering = true
	}

	for k, vv := range *src {
		if x, ok := f[k]; filtering && (!ok || !x) {
			continue
		}
		for _, v := range vv {
			dst.Add(k, v)
		}
	}
}

// sets a proxy metrics (ProxyMetrics interface) for the proxy
func (p *Proxy) SetMetricsCollector(pm ProxyMetrics) {
	p.metrics = pm
}

// Returns a new Proxy. An error is returned if there was a failure
// to parse the regex from the passed Config.
func New(pc Config) (*Proxy, error) {
	tr := &http.Transport{
		Dial: func(netw, addr string) (net.Conn, error) {
			c, err := net.DialTimeout(netw, addr, pc.RequestTimeout)
			if err != nil {
				return nil, err
			}
			// also set time limit on reading
			c.SetDeadline(time.Now().Add(pc.RequestTimeout))
			return c, nil
		}}

	// spawn an idle conn trimmer
	go func() {
		// prunes every 5 minutes. this is just a guess at an
		// initial value. very busy severs may want to lower this...
		time.Sleep(5 * time.Minute)
		tr.CloseIdleConnections()
	}()

	// build/compile regex
	client := &http.Client{Transport: tr}
	if pc.NoFollowRedirects {
		client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
			return errors.New("Not following redirect")
		}
	}

	allow := make([]*regexp.Regexp, 0)
	var c *regexp.Regexp
	var err error
	// compile allow list
	for _, v := range pc.AllowList {
		c, err = regexp.Compile(v)
		if err != nil {
			return nil, err
		}
		allow = append(allow, c)
	}

	return &Proxy{
		client:    client,
		hmacKey:   []byte(pc.HmacKey),
		allowList: allow,
		maxSize:   pc.MaxSize}, nil
}


camoproxy/encoding/url.go
package encoding

import (
	"crypto/hmac"
	"crypto/sha1"
	"crypto/subtle"
	"encoding/base64"
	"encoding/hex"
	"github.com/cactus/gologit"
	"strings"
)

type sEncoder func([]byte) string
type sDecoder func(string) ([]byte, error)
type Encoder func(*[]byte, string) string
type Decoder func(*[]byte, string, string) (string, bool)

// DecodeUrl ensures the url is properly verified via HMAC, and then
// decodes the hex url, returning the url (if valid) and whether the
// HMAC was verified.
func DecodeHexUrl(hmackey *[]byte, encdig string, encurl string) (string, bool) {
	decoder := hex.DecodeString
	return decodeUrl(hmackey, decoder, encdig, encurl)
}

// DecodeUrl ensures the url is properly verified via HMAC, and then
// decodes the base64 url, returning the url (if valid) and whether the
// HMAC was verified.
func DecodeBase64Url(hmackey *[]byte, encdig string, encurl string) (string, bool) {
	decoder := func(s string) ([]byte, error) {
		if len(s) % 4 != 0 {
			s = s + strings.Repeat("=", (4 - (len(s) % 4)))
		}
		return base64.URLEncoding.DecodeString(s)
	}
	return decodeUrl(hmackey, decoder, encdig, encurl)
}

// DecodeUrl ensures the url is properly verified via HMAC,
// decodes the url with the provided decoder, returning the url (if valid) and
// whether the HMAC was verified.
func decodeUrl(hmackey *[]byte, decoder sDecoder, encdig string, encurl string) (surl string, valid bool) {
	urlBytes, err := decoder(encurl)
	if err != nil {
		gologit.Debugln("Bad URL decode", encurl)
		return
	}
	inMacSum, err := decoder(encdig)
	if err != nil {
		gologit.Debugln("Bad digest decode", encdig, err)
		return
	}

	mac := hmac.New(sha1.New, *hmackey)
	mac.Write(urlBytes)
	macSum := mac.Sum(nil)

	if subtle.ConstantTimeCompare(macSum, inMacSum) != 1 {
		gologit.Debugf("Bad signature: %x != %x\n", macSum, inMacSum)
		return
	}
	surl = string(urlBytes)
	valid = true
	return
}

// EncodeHexUrl takes an HMAC key and a url, and returns url
// path partial consisitent of signature and hex encoded url.
func EncodeHexUrl(hmacKey *[]byte, oUrl string) string {
	encoder := hex.EncodeToString
	return encodeUrl(hmacKey, encoder, oUrl)
}

// EncodeHexUrl takes an HMAC key and a url, and returns url
// path partial consisitent of signature and base64 encoded url.
func EncodeBase64Url(hmacKey *[]byte, oUrl string) string {
	encoder := func(src []byte) string {
		return strings.TrimRight(base64.URLEncoding.EncodeToString(src), "=")
	}
	return encodeUrl(hmacKey, encoder, oUrl)
}

// encodeUrl takes an HMAC key, an encoder function, and a url, and returns url
// path partial consisitent of signature and encoded url.
func encodeUrl(hmacKey *[]byte, encoder sEncoder, oUrl string) string {
	mac := hmac.New(sha1.New, *hmacKey)
	mac.Write([]byte(oUrl))
	macSum := encoder(mac.Sum([]byte{}))
	encodedUrl := encoder([]byte(oUrl))
	encurl := "/" + macSum + "/" + encodedUrl
	return encurl
}


src/go-camo/camoproxy/stats.go
package camoproxy

import (
	metrics "github.com/rcrowley/go-metrics"
	"bytes"
	"fmt"
)

type proxyStats struct {
	registry	  *metrics.StandardRegistry
	Enable        bool
}

func NewProxyStats() *proxyStats {
	r := metrics.NewRegistry()
	c := metrics.NewMeter()
	r.Register("clients", c)
	c.Mark(0)
	b := metrics.NewMeter()
	r.Register("bytes", b)
	b.Mark(0)
	ps := &proxyStats{registry: r}
	ps.Enable = true
	return ps
}

func (ps *proxyStats) AddServed() {
	if ps.Enable != true {
		return
	}
	m := ps.registry.Get("clients")
	if m == nil {
		return
	}
	m.(metrics.Meter).Mark(1)
}

func (ps *proxyStats) AddBytes(bc int64) {
	if ps.Enable != true {
		return
	}
	if bc <= 0 {
		return
	}
	m := ps.registry.Get("bytes")
	if m == nil {
		return
	}
	m.(metrics.Meter).Mark(bc)
}

func (ps *proxyStats) GetStats() *bytes.Buffer {
	buf := &bytes.Buffer{}
	for _, stat := range []string{"clients", "bytes"} {
		m := ps.registry.Get(stat).(metrics.Meter)
		if m == nil {
			continue
		}
		// mark values so they update
		m.Mark(0)
		fmt.Fprintf(buf, "meter %s\n", stat)
		fmt.Fprintf(buf, "  count:       %9d\n", m.Count())
		fmt.Fprintf(buf, "  1-min rate:  %12.2f\n", m.Rate1())
		fmt.Fprintf(buf, "  5-min rate:  %12.2f\n", m.Rate5())
		fmt.Fprintf(buf, "  15-min rate: %12.2f\n", m.Rate15())
		fmt.Fprintf(buf, "  mean rate:   %12.2f\n", m.RateMean())
	}
	return buf
}


go-camo.go
// go-camo daemon (go-camod)
package main

import (
	"code.google.com/p/gorilla/mux"
	"encoding/json"
	"fmt"
	"github.com/cactus/go-camo/camoproxy"
	"github.com/cactus/gologit"
	flags "github.com/jessevdk/go-flags"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"runtime"
	"strconv"
	"syscall"
	"time"
)

func main() {
	var gmx int
	if gmxEnv := os.Getenv("GOMAXPROCS"); gmxEnv != "" {
		gmx, _ = strconv.Atoi(gmxEnv)
	} else {
		gmx = runtime.NumCPU()
	}
	runtime.GOMAXPROCS(gmx)

	// command line flags
	var opts struct {
		ConfigFile     string        `short:"c" long:"config" description:"JSON Config File"`
		HmacKey        string        `short:"k" long:"key" description:"HMAC key"`
		Stats          bool          `long:"stats" description:"Enable Stats"`
		MaxSize        int64         `long:"max-size" default:"5120" description:"Max response image size (KB)"`
		ReqTimeout     time.Duration `long:"timeout" default:"4s" description:"Upstream request timeout"`
		NoFollow       bool          `long:"no-follow" description:"Disable following upstream redirects"`
		BindAddress    string        `long:"listen" default:"0.0.0.0:8080" description:"Address:Port to bind to for HTTP"`
		BindAddressSSL string        `long:"ssl-listen" description:"Address:Port to bind to for HTTPS/SSL/TLS"`
		SSLKey         string        `long:"ssl-key" description:"ssl private key (key.pem) path"`
		SSLCert        string        `long:"ssl-cert" description:"ssl cert (cert.pem) path"`
		Verbose        bool          `short:"v" long:"verbose" description:"Show verbose (debug) log level output"`
		Version        bool          `short:"V" long:"version" description:"print version and exit"`
	}

	// parse said flags
	_, err := flags.Parse(&opts)
	if err != nil {
		if e, ok := err.(*flags.Error); ok {
			if e.Type == flags.ErrHelp {
				os.Exit(0)
			}
		}
		os.Exit(1)
	}

	if opts.Version {
		fmt.Println(camoproxy.ServerNameVer)
		os.Exit(0)
	}

	config := camoproxy.Config{}

	if opts.ConfigFile != "" {
		b, err := ioutil.ReadFile(opts.ConfigFile)
		if err != nil {
			log.Fatal("Could not read configFile", err)
		}
		err = json.Unmarshal(b, &config)
		if err != nil {
			log.Fatal("Could not parse configFile", err)
		}
	}

	// flags override config file
	if opts.HmacKey != "" {
		config.HmacKey = opts.HmacKey
	}

	if config.MaxSize == 0 {
		config.MaxSize = opts.MaxSize
	}

	if opts.BindAddress == "" && opts.BindAddressSSL == "" {
		log.Fatal("One of bind-address or bind-ssl-address required")
	}

	if opts.BindAddressSSL != "" && opts.SSLKey == "" {
		log.Fatal("ssl-key is required when specifying bind-ssl-address")
	}
	if opts.BindAddressSSL != "" && opts.SSLCert == "" {
		log.Fatal("ssl-cert is required when specifying bind-ssl-address")
	}

	// convert from KB to Bytes
	config.MaxSize = config.MaxSize * 1024
	config.RequestTimeout = opts.ReqTimeout
	config.NoFollowRedirects = opts.NoFollow

	// set logger debug level and start toggle on signal handler
	logger := gologit.Logger
	logger.Set(opts.Verbose)
	logger.Debugln("Debug logging enabled")
	logger.ToggleOnSignal(syscall.SIGUSR1)

	proxy, err := camoproxy.New(config)
	if err != nil {
		log.Fatal(err)
	}

	router := mux.NewRouter()
	router.Handle("/favicon.ico", http.NotFoundHandler())
	// hex url
	router.Handle("/{sigHash}/{encodedUrl}", proxy).Methods("GET")
	router.HandleFunc("/", RootHandler)
	http.Handle("/", router)

	if opts.Stats {
		ps := &ProxyStats{}
		proxy.SetMetricsCollector(ps)
		log.Println("Enabling stats at /status")
		router.Handle("/status", StatsHandler(ps))
	}

	if opts.BindAddress != "" {
		log.Println("Starting server on", opts.BindAddress)
		go func() {
			log.Fatal(http.ListenAndServe(opts.BindAddress, nil))
		}()
	}
	if opts.BindAddressSSL != "" {
		log.Println("Starting TLS server on", opts.BindAddressSSL)
		go func() {
			log.Fatal(http.ListenAndServeTLS(
				opts.BindAddressSSL, opts.SSLCert, opts.SSLKey, nil))
		}()
	}

	// just block. listen and serve will exit the program if they fail/return
	// so we just need to block to prevent main from exiting.
	select {}
}
